src/main/java/com/cloudbees/jenkins/support/impl/LoadStats.java

```diff
@@ -0,0 +1,180 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.jenkins.support.impl;
+
+import com.cloudbees.jenkins.support.api.Component;
+import com.cloudbees.jenkins.support.api.Container;
+import com.cloudbees.jenkins.support.api.Content;
+import com.cloudbees.jenkins.support.api.PrintedContent;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Extension;
+import hudson.model.Label;
+import hudson.model.LoadStatistics;
+import hudson.model.MultiStageTimeSeries;
+import hudson.model.TimeSeries;
+import hudson.security.Permission;
+import jenkins.model.Jenkins;
+
+import javax.imageio.ImageIO;
+import java.awt.*;
+import java.awt.image.BufferedImage;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.io.UnsupportedEncodingException;
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.net.URLEncoder;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TimeZone;
+import java.util.TreeMap;
+
+/**
+ * @since 
+ */
+@Extension
+public class LoadStats extends Component {
+    @NonNull
+    @Override
+    public Set<Permission> getRequiredPermissions() {
+        return Collections.emptySet();
+    }
+
+    @NonNull
+    @Override
+    public String getDisplayName() {
+        return "Load Statistics";
+    }
+
+    @Override
+    public void addContents(@NonNull Container container) {
+        Jenkins jenkins = Jenkins.getInstance();
+        add(container, "unlabeled", jenkins.unlabeledLoad);
+        add(container, "overall", jenkins.overallLoad);
+        for (Label l: jenkins.getLabels()) {
+            try {
+                add(container, String.format("label/%s", URLEncoder.encode(l.getName(), "UTF-8")), l.loadStatistics);
+            } catch (UnsupportedEncodingException e) {
+                // ignore UTF-8 is required by JLS specification
+            }
+        }
+    }
+
+    private void add(@NonNull Container container, String name, LoadStatistics stats) {
+        boolean headless = GraphicsEnvironment.isHeadless();
+        for (MultiStageTimeSeries.TimeScale scale: MultiStageTimeSeries.TimeScale.values()) {
+            if (!headless) {
```

Comment:
For me, the graphs are a nice to have, it's the raw data that matters. What
I may do is output a gnu plot script to generate the PNGs if running
headless... Though that doesn't add much value as people consuming support
bundles should be able to have that script on standby... It would compress
well and restore value to sysadmins trying to self-diagnose

On Saturday, April 25, 2015, Daniel Beck <notifications@github.com> wrote:

> In src/main/java/com/cloudbees/jenkins/support/impl/LoadStats.java
> <https://github.com/jenkinsci/support-core-plugin/pull/29#discussion_r29099473>
> :
>
> > +        Jenkins jenkins = Jenkins.getInstance();
> > +        add(container, "unlabeled", jenkins.unlabeledLoad);
> > +        add(container, "overall", jenkins.overallLoad);
> > +        for (Label l: jenkins.getLabels()) {
> > +            try {
> > +                add(container, String.format("label/%s", URLEncoder.encode(l.getName(), "UTF-8")), l.loadStatistics);
> > +            } catch (UnsupportedEncodingException e) {
> > +                // ignore UTF-8 is required by JLS specification
> > +            }
> > +        }
> > +    }
> > +
> > +    private void add(@NonNull Container container, String name, LoadStatistics stats) {
> > +        boolean headless = GraphicsEnvironment.isHeadless();
> > +        for (MultiStageTimeSeries.TimeScale scale: MultiStageTimeSeries.TimeScale.values()) {
> > +            if (!headless) {
>
> JENKINS-25261 <https://issues.jenkins-ci.org/browse/JENKINS-25261> has
> some related discussion. At the same time, RPM installs
> <https://github.com/jenkinsci/packaging/blob/master/rpm/build/SOURCES/jenkins.sysconfig.in#L37>
> default to headless (but can generate graphs if there are fonts).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/support-core-plugin/pull/29/files#r29099473>
> .
>


-- 
Sent from my phone